Add BlockPool allocator for GCNode allocation

llbit opened the pull request:
This replaces the use of BDWGC for allocating GCNodes. The BlockPool
allocator allocates small objects in an arena, this speeds up allocation
and pointer lookup for scalars.

llbit commented:
Should BlockPool be in the rho namespace?

kmillar commented:
Yes

On Thu, Jul 14, 2016 at 4:15 PM, Jesper Öqvist <notifications@github.com>
wrote:

> Should BlockPool be in the rho namespace?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rho-devel/rho/pull/106#issuecomment-232819509>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFPJXvAyAxhRTLNfpq1HaBVKEgwxyN3oks5qVsMJgaJpZM4JM1yi>
> .
>

kmillar commented on src/main/BlockPool.cpp:
```diff
@@ -0,0 +1,863 @@
+using std::size_t;
+
+// Tracking heap bounds for fast rejection in pointer lookup.
+void* heap_start = reinterpret_cast<void*>(UINTPTR_MAX);
```
Variables and function that aren't exported should be declared 'static' or in an anonymous namespace.

kmillar commented on src/main/BlockPool.cpp:
```diff
@@ -0,0 +1,863 @@
+
+// Free a pointer inside a given superblock. The block MUST be in the given
+// superblock.
+void free_small_block(void* pointer, Superblock* superblock) {
```
Is there any reason why this isn't a method in Superblock?  (and similarly for some of the other functions)

llbit replied:
No particular reason, I'll move this to Superblock.

kmillar commented on src/main/BlockPool.cpp:
```diff
@@ -0,0 +1,863 @@
+
+// Add an allocation (large object or medium object superblock) to the
+// sparse allocations hashtable.
+void add_sparse_block(uintptr_t pointer, size_t size) {
```
What does 'sparse' mean in this context?  Can it be renamed to make it clearer?

kmillar commented on src/main/BlockPool.cpp:
```diff
@@ -0,0 +1,863 @@
+
+unsigned sparse_bits = 16;
+unsigned num_sparse_buckets = 1 << sparse_bits;
+unsigned hash_mask = num_sparse_buckets - 1;
```
Can the sparse hash table be turned into a class?

kmillar commented on src/main/BlockPool.cpp:
```diff
@@ -0,0 +1,863 @@
+
+// Computes the hash key for a pointer.
+static unsigned hash_ptr(uintptr_t ptr, unsigned hash_mask) {
+  unsigned low = (ptr << 4) & 0xFFFFFFFF;
```
Do you really mean '<<' here, not '>>'?

llbit replied:
Yes, the argument has already been right-shifted. This left-shift is used to spread out the hash keys, so that adjacent objects are not sequential in the hash table. This leaves some room for collision buckets between the hash keys.

kmillar commented on src/main/BlockPool.cpp:
```diff
@@ -0,0 +1,863 @@
+#define MAX_COLLISIONS (6)
+#define MAX_REBALANCE_COLLISIONS (3)
+
+// Computes the hash key for a pointer.
```
Please add some comments or a reference describing why this hash function works the way it does.

llbit replied:
Sure, will do!

kmillar commented on src/main/GCNode.cpp:
Can you implement this functionality in your allocator as well when asan is enabled?  This is useful for finding bugs.

llbit commented on src/main/GCNode.cpp:
Yes, that would be possible to add!

kmillar commented on src/main/GCNode.cpp:
```diff
@@ -262,121 +199,81 @@ void GCNode::mark()
+    vector<void*> work;
+    vector<GCNode*> to_delete;
+    BlockPool::ApplyToAllBlocks([&](void* pointer) {
+        work.push_back(pointer);
```
Why do we need to create this vector with all the pointers?  IIUC, it going to contain every currently allocated pointer, which could be huge.

llbit replied:
The worklist was used to avoid problems caused by deleting nodes while iterating over live nodes. This is not a problem in the current version of the allocator since it does not use an std::map or other datastructure that disallows concurrent iteration/modification.

The worklist could be removed right now. The only pitfall is if the bitset is not volatile, in which case the allocator could mistakenly identify something as allocated. I had this problem in yet another previous version of the allocator.

llbit replied:
I have deleted the worklist since it is not currently needed, and apparently we don't need volatile, just need to not copy the bitsets to locals.

kmillar commented on src/include/rho/BlockPool.hpp:
```diff
@@ -0,0 +1,57 @@
+using std::function;
+
+namespace rho {
+  class BlockPool {
```
The word 'block' gets used in a bunch of different ways in the allocator.  It would be clearer to call this an 'Allocator', with methods Allocate(), Free() and ApplyToAllAllocations().

kmillar replied:
Add a comment explaining the purpose of this class.  In particular, it should be noted that only GCNode objects are stored using this allocator.

llbit replied:
@kmillar "Allocator" seems to collide with another class already in the `rho` namespace. What about `GCNodeAllocator`?

kmillar replied:
SGTM

kmillar commented on src/main/BlockPool.cpp:
```diff
@@ -0,0 +1,1216 @@
+ * and 1040 * 8 * 64 = 532480 < 2^19 - 1168. So all objects are mappable
+ * for a 2^19 superblock.
+ */
+#define SUPERBLOCK_HEADER_SIZE (1168)
```
Can this be expressed as a calculation, instead of a fixed constant?  Otherwise, it'd need manually updating if we change anything.

kmillar commented on src/main/BlockPool.cpp:
```diff
@@ -0,0 +1,1216 @@
+#include "rho/BlockPool.hpp"
+
+#ifdef HAVE_ADDRESS_SANITIZER
+#include "rho/AddressSanitizer.hpp"
```
This should be unconditionally included, as it's responsible for detecting if asan is being used and setting HAVE_ADDRESS_SANITIZER

kmillar commented on src/main/BlockPool.cpp:
This won't compile with HAVE_ADDRESS_SANITIZER, as FreeListNode hasn't been declared yet.

llbit commented on src/main/BlockPool.cpp:
This was a silly mistake due to reorganizing and not building with Address Sanitizer after. Thanks for catching this!

kmillar commented on src/main/BlockPool.cpp:
Make this a non-static member acting on superblock

kmillar commented on src/main/BlockPool.cpp:
Make this a non-static member acting on superblock
 Add a line note

kmillar commented on src/main/BlockPool.cpp:
This should be a quick function that then dispatches to Superblock::AllocateSmallBlock()

kmillar commented on src/main/BlockPool.cpp:
The hashtable uses open addressing with quadratic probing to resolve collisions.

kmillar commented on src/main/BlockPool.cpp:
Document how it differs from dense_hash_table and why.

kmillar commented on src/main/BlockPool.cpp:
This should be private, and probably should be renamed 'resize'.

kmillar commented on src/main/BlockPool.cpp:
Is there any reason why we can't easily support 64 bits here?

kmillar commented on src/main/BlockPool.cpp:
```diff
@@ -0,0 +1,1216 @@
+/*
```
This file has gotten quite large and complex.  Please split it into multiple files.

Allocator.*
AllocatorSuperblock.*
AllocatorTable.*
etc.

Thanks

llbit replied:
Done!

llbit commented on src/main/BlockPool.cpp:
Making the `AddressSanitizer.hpp` include unconditional and building with Address Sanitizer uncovered a couple errors which I am fixing now. Some parts of the Address Sanitizer support were definitely working because I was getting ASAN runtime errors previously, but then I made the include conditional at some point and assumed stuff was working. D'oh!

llbit commented on src/main/BlockPool.cpp:
The hash key is usable up to 64 bits. This limit is a leftover from when my code was flaky and would eat up enormous amounts of memory and crash my machine fairly regularly. I put this limit in here so that the program would halt instead of eating all memory in the universe.

I will remove it!

llbit commented on src/main/BlockPool.cpp:
Although, if the hash key size is 64 bits then the hashtable has 2^64 entries, so it's unreasonable to go that high.

kmillar commented on src/include/rho/AllocationTable.hpp:
```diff
@@ -0,0 +1,173 @@
+    void PrintTable();
+
+  private:
+    unsigned num_bits;
```
Prefix the names of class members with 'm_'

llbit replied:
Done.

kmillar commented on src/include/rho/AllocationTable.hpp:
```diff
@@ -0,0 +1,173 @@
+     */
+    struct HashBucket {
+      uintptr_t data;
+      unsigned size;
```
Should be size_t.  We support vectors with sizes > 4GB.

llbit replied:
The actual size is 2^size for large objects.

kmillar replied:
Please rename it to make it clear then.

llbit replied:
Okay, will do.

kmillar commented on src/include/rho/AllocationTable.hpp:
```diff
@@ -0,0 +1,173 @@
+typedef std::uint32_t u32;
+
+using std::function;
+using std::size_t;
```
Please don't add 'typedef' and 'using' declarations in header files as they pollute the global namespace.  They're fine in .cc files.

kmillar commented on src/include/rho/AllocationTable.hpp:
```diff
@@ -0,0 +1,173 @@
+     * if the hash bucket contains a superblock (2) and whether this
+     * hash bucket is the first representing that allocation (1).
+     */
+    struct HashBucket {
```
Needs a better name.  Maybe Allocation.

llbit replied:
Ok, renamed to Allocation.

kmillar commented on src/include/rho/AllocationTable.hpp:
```diff
@@ -0,0 +1,173 @@
+     * Hash bucket for the allocation table.
+     * The two lowest bits of the data pointer are flags indicating
+     * if the hash bucket contains a superblock (2) and whether this
+     * hash bucket is the first representing that allocation (1).
```
Add a constructor and some getter functions that encapsulate this.

llbit replied:
Done!

kmillar commented on src/include/rho/AllocationTable.hpp:
```diff
@@ -0,0 +1,173 @@
+    HashBucket* buckets = nullptr;
+
+    // Counters logging collision statistics:
+    unsigned add_collisions = 0;
```
These should be insert / erase / search to match the names of the functions.  Adding 'num_' would make it clearer too.
    m_num_insert_collisions

llbit replied:
Done and done.

kmillar commented on src/main/AllocationTable.cpp:
```diff
@@ -0,0 +1,334 @@
+
+// The low pointer bits are masked out in the hash function.
+#define LOW_BITS (19)
+#define MAX_COLLISIONS (6)
```
When the table gets large, it's going to need to be really sparse to satisfy this.  With n elements in a hash table of size m, the expected number of collisions ~ n^2 / m, so we'd need m ~ n^2.

Reasonable options would be to either make the number of allowed collisions proportional to the size of the table, or just resize based on the load factor.

llbit replied:
I am looking into this. Thanks.

kmillar commented on src/main/AllocationTable.cpp:
```diff
@@ -0,0 +1,334 @@
+  uintptr_t key_end = (pointer + (1 << size) + ((1 << LOW_BITS) - 1))
+      >> LOW_BITS;
+  bool first = true;
+  do {
```
For clarity, please implement this as a for loop, and add a comment.

llbit replied:
It gets awkward with a for loop, because this is one of those rare occasions when the loop should always take one iteration even if the first key is equal to the end key. If the first key is not equal to the end key then the end key should not be inserted.

kmillar commented on src/include/rho/AllocatorSuperblock.hpp:
```diff
@@ -0,0 +1,156 @@
+ * A small object superblock has size 2^18.
+ * Minimum object size = 32 bytes.
+ */
+#define SUPERBLOCK_BITS (18)
```
Instead of using #define, you can do something like:
  static constexpr unsigned SUPERBLOCK_BITS = 18;
etc. inside the class scope.

kmillar commented on src/include/rho/AllocationTable.hpp:
```diff
@@ -0,0 +1,173 @@
+    bool free_allocation(uintptr_t pointer);
+
+    /**
+     * Iterates over all live objects and calls the argument function on
```
I think this iterates over all allocated objects, not just the live ones.

kmillar commented on src/include/rho/GCNodeAllocator.hpp:
I think this applies to all current allocations, not just the live ones.

kmillar commented on src/include/rho/GCNodeAllocator.hpp:
std::function<void(GCNode*)> f

kmillar commented on src/include/rho/GCNodeAllocator.hpp:
Should return a GCNode*

llbit commented on src/include/rho/GCNodeAllocator.hpp:
What do you mean by "live" allocation? A node that is still reachable from a root?

What I meant by "live" was a GCNode that has not yet been freed yet, (so potentially unreachable just not collected yet).

kmillar commented on src/main/AllocationTable.cpp:
```diff
@@ -0,0 +1,334 @@
+          }
+        }
+      } else if (pointer <= candidate
+            && (pointer + (1L << bucket->size)) > candidate) {
```
The calculation of the end of the memory in the allocation should be a HashBucket / Allocation method.

kmillar commented on src/include/rho/AllocationTable.hpp:
```diff
@@ -0,0 +1,250 @@
+     * hash bucket is the first representing that allocation (1).
+     */
+    class Allocation {
+    public:
```
Please make the relevant methods and members private.

llbit replied:
I've made most of the members private now.

kmillar commented on src/include/rho/AllocationTable.hpp:
```diff
@@ -0,0 +1,326 @@
+     * superblock, the superblock is left as is but the superblock bitset is
+     * updated to indicate that the block is free.
+     * If the allocation is a large object (ie. not in a superblock), then
+     * the object is taken out of the hashtable and inserted into a freelist.
```
All of this stuff about superblocks etc doesn't really belong in the AllocationTable, but rather in the code that uses it.  

Presumably you'll need to add something to the api to let the caller know if an allocation is a superblock or not.

llbit replied:
Yes, the freeAllocation function could be moved out of AllocationTable. The reason I left it in AllocationTable was to avoid adding a dependency from the caller on AllocationTable::Allocation (which knows if the contained allocation is a superblock) and I wanted AllocationTable::Allocation to only be used inside AllocationTable.

kmillar commented on src/main/AllocationTable.cpp:
```diff
@@ -0,0 +1,286 @@
+}
+
+void rho::AllocationTable::insert(void* pointer, std::size_t size) {
+  if (m_capacity == 0) {
```
You may need more capacity than this, since you're doing multiple inserts.

llbit replied:
Thanks for spotting this! I now use the start and end keys to calculate the required capacity for each allocation.

kmillar commented on src/include/rho/AllocationTable.hpp:
```diff
@@ -0,0 +1,326 @@
+     * because it is used to tentatively insert new nodes into a temporary
+     * resized table.
+     *
+     * @return true if the insert succeeded.
```
With the probe sequence that you're using, insertion should only fail if the table is completely full. Make it a precondition that the table has enough space, and you can remove the failure case, which will make things simpler.

llbit replied:
This made things much simpler!

kmillar commented:
Please add your unit tests under tests/rho

kmillar commented on src/main/AllocationTable.cpp:
```diff
@@ -0,0 +1,286 @@
+  for (size_t key = startKey(pointer); key < key_end; ++key) {
+    size_t hash = pointerHash(key << LOW_BITS);
+    bool added = false;
+    for (int i = 0; i < m_num_buckets; ++i) {
```
Split the inner loop into its own function.

llbit replied:
Done.

kmillar commented on src/main/AllocationTable.cpp:
```diff
@@ -0,0 +1,286 @@
+    // Ensure that at least one entry is added to the table.
+    // If the pointer was exactly aligned to LOW_BITS then
+    // key_end will be equal to key, otherwise key_end will point
+    // to the next hash bucket past the allocation range.
```
I don't see this.  with size = 0:
   key_end = startKey(pointer) + (1 + (1 << LOW_BITS) - 1) >> LOW_BITS
                  = startKey(pointer) + 1

llbit replied:
Ah, you are right. Since the allocation size is never 0 it always pushes the end key to the next hash key regardless of whether the allocation was aligned, so I can just delete that check!

kmillar commented on src/main/AllocationTable.cpp:
```diff
@@ -0,0 +1,286 @@
+  bool first = true;
+  for (size_t key = startKey(pointer); key < key_end; ++key) {
+    size_t hash = pointerHash(key << LOW_BITS);
+    for (int i = 0; i < m_num_buckets; ++i) {
```
Split the inner loop into its own function

llbit replied:
Done.

kmillar commented on src/main/AllocationTable.cpp:
```diff
@@ -0,0 +1,286 @@
+        break;
+      }
+      if (bucket.isEmpty()) {
+        // No more partial entries inserted, we are done.
```
If we get here, then we're erasing something that we never inserted into the table.  Please change this to a fatal error.

llbit replied:
Yes, this behavior was left over from when I expected every insert to be able to fail, and erase was used to clean up partially inserted keys. Made it more sane now so that it generates an abort().

kmillar commented on src/main/AllocationTable.cpp:
```diff
@@ -0,0 +1,286 @@
+  size_t hash = pointerHash(candidate);
+  for (int i = 0; i < m_num_buckets; ++i) {
+    Allocation* bucket = &m_buckets[hash];
+    if (!bucket->isEmpty() && !bucket->isDeleted()) {
```
You can make the code clearer by reorganizing the conditions:
```
  if (bucket->isEmpty())
           return nullptr;
  if (bucket->isDeleted()) {
         m_num_search_collisions += 1;
         hash = probeFunc(hash, i);
         continue;
   }
   
   if (bucket->isSuperblock()) {
    ....
```

llbit replied:
Thanks, I never thought of having the probe function call in the loop update!

kmillar commented on src/main/AllocationTable.cpp:
```diff
@@ -0,0 +1,286 @@
+      // This is the initial hash bucket for the corresponding allocation.
+      // Rebuild the hash entries in the new table based only on this entry.
+      new_capacity -= 1;
+      if (!tryInsert(new_table, bucket.m_data & ~uintptr_t{1},
```
Since tryInsert never needs to fail, this entire function can be simplified a bunch.

If that wasn't the case, you could simplify as:
    AllocationTable* new_table = new AllocationTable(new_allocatable_bits);
    ... copy over the entries or delete new_table and exit.
    std::swap(*this, *new_table);

kmillar commented on src/include/rho/AllocationTable.hpp:
```diff
@@ -0,0 +1,326 @@
+    /**
+     * \brief Frees the memory used by the hashtable, but not the allocations in it.
+     */
+    ~AllocationTable();
```
You also need to delete (or implement) the copy and assignment operators.

kmillar commented on src/main/GCNodeAllocator.cpp:
You can completely unroll this loop if you have the last element in the quarantine (or loop to find it):

```
last_element_in_quarantine->m_next = s_freelists[i];
s_freelists[i] = s_quarantine[i];
s_quarantine[i] = nullptr;
```

kmillar commented on src/main/GCNodeAllocator.cpp:
Move this into a separate function

kmillar commented on src/include/rho/GCNodeAllocator.hpp:
```diff
@@ -0,0 +1,223 @@
+   * Freelists are used to store superblock nodes and large allocations.
+   * For large allocations the block and superblock members are unused.
+   */
+  struct FreeListNode {
```
Consider using boost::intrusive::slist instead.  See GCStackFrameBoundary for an example.

llbit replied:
After looking into this more, I think using `boost::intrusive::slist` is difficult to integrate well with the quarantine list because it needs to unpoison the allocations in the freelist during iteration.

kmillar commented on src/include/rho/GCNodeAllocator.hpp:
```diff
@@ -0,0 +1,223 @@
+      static AllocationTable* s_alloctable;
+
+      /** Number of different object sizes plus one for the small-object superblocks. */
+      static constexpr unsigned s_num_small_pools = (256 / 8) + 1;
```
Use parameters, not constants in here and throughout.  e.g.:
    = s_max_small_block_size / s_small_block_spacing + 1;

kmillar commented on src/include/rho/GCNodeAllocator.hpp:
```diff
@@ -0,0 +1,223 @@
+      /**
+       * \brief Add a free list node to a free list by size class.
+       */
+      static void addToFreelist(FreeListNode* free_node, unsigned size_class);
```
Having two functions with the same name, but different semantics (at least for superblocks) is error-prone.

llbit replied:
I agree, this was a poor choice. I renamed the first one to differentiate the two.

kmillar commented on src/include/rho/GCNodeAllocator.hpp:
```diff
@@ -0,0 +1,223 @@
+      /**
+       * Given a size class > s_num_small_pools, return the log2 of the size.
+       */
+      static unsigned sizeLog2FromSizeClass(unsigned size_class) {
```
Add assert statements to check the preconditions.

llbit replied:
Done.

kmillar commented on src/main/AllocatorSuperblock.cpp:
```diff
@@ -0,0 +1,297 @@
+#include "rho/AllocationTable.hpp"
+#include "rho/AllocatorSuperblock.hpp"
+
+unsigned rho::AllocatorSuperblock::s_size_class[] = {
```
Isn't this the same as bytesFromSizeClass() ?

llbit replied:
Yes, I was planning to delete that and just use the function, but I forgot to delete it.

kmillar commented on src/main/AllocatorSuperblock.cpp:
```diff
@@ -0,0 +1,297 @@
+
+void rho::AllocatorSuperblock::allocateArena() {
+  // We use new[] here, but one could also use sbrk(ARENASIZE).
+  superblock_arena = new double[s_arenasize / sizeof(double)];
```
It'd be clearer to use 'new char[s_arenasize]' or just malloc(s_arenasize)

llbit replied:
Rewrote this bit.

kmillar commented on src/main/AllocatorSuperblock.cpp:
```diff
@@ -0,0 +1,297 @@
+  superblock_arena = new double[s_arenasize / sizeof(double)];
+  uintptr_t start = reinterpret_cast<uintptr_t>(superblock_arena);
+  uintptr_t end = start + s_arenasize;
+  uintptr_t pad = s_small_superblock_size - (start & (s_small_superblock_size - 1));
```
Actually, you probably should be using posix_memalign.  Failing that, std::align can convert superblock_arena to arena_superblock_start.

llbit replied:
For some reason, `std::align` is missing, but `posix_memalign` works fine for me.

kmillar commented on src/main/AllocatorSuperblock.cpp:
```diff
@@ -0,0 +1,297 @@
+};
+
+// Arena for small-object superblocks:
+void* superblock_arena = nullptr;
```
AFAICT, this isn't used anywhere outside of allocateArena(), so it can be turned into a local variable.

llbit replied:
Done.

kmillar commented on src/main/GCNodeAllocator.cpp:
Either the variable name or the comment should be changed, so they match.

llbit commented on src/main/GCNodeAllocator.cpp:
That's clever! Thanks!

kmillar commented on src/main/GCNodeAllocator.cpp:
is this the same as `if (size > (1 << log2)`?

llbit commented on src/main/GCNodeAllocator.cpp:
Yes, no idea why I made it so complicated.

llbit commented on src/main/GCNodeAllocator.cpp:
Thanks, that's an outdated comment I forgot to update.

llbit commented on src/main/GCNodeAllocator.cpp:
Donesies.

llbit commented:
@kmillar Thank you for all the comments so far! I added unit tests and addressed your comments in the latest commit.

I decided not to use `boost::intrusive::slist` yet, because I did not find a way to integrate it with the quarantine freelist. Do you have an idea I might get around the Address Sanitizer poisoned freelist? The only idea I have is that the intrusive list part of the freelist could be kept unpoisoned.

ArunChauhan commented:
Your code uses the "const" qualifier on methods (which is good), but it is not used consistently.  I suggest going over your code once to mark all the const methods.  It helps the compiler optimize the methods.

ArunChauhan commented on src/main/GCNodeAllocator.cpp:
If you adopted this algorithm from a third-party source then it may be a good idea to cite it here.

ArunChauhan commented on src/main/GCNodeAllocator.cpp:
Nit: The C++ idiom to get file-scoped declarations is to enclose them in anonymous namespace, instead of declaring them static.

llbit commented on src/main/GCNodeAllocator.cpp:
The algorithm was inspired by the similar but opposite operation of finding the number of trailing zeroes by binary search here: https://graphics.stanford.edu/~seander/bithacks.html#ZerosOnRightBinSearch

I added a reference in a comment.

llbit commented on src/main/GCNodeAllocator.cpp:
OK, wrapped these in an anonymous namespace.

llbit commented:
@kmillar I can rebase this into one commit if it helps. I avoided doing that so far so that the old comments would not disappear.

kmillar commented on src/main/AllocatorSuperblock.cpp:
```diff
@@ -0,0 +1,307 @@
+  void* arena = nullptr;
+  size_t space = s_arenasize; // Total acquired arena space.
+  posix_memalign(&arena, s_small_superblock_size, space);
+#ifdef HAVE_STD_ALING
```
This macro is never defined (and needs s/NG/GN/).  In any case, unless s_arenasize gets an invalid value, posix_memalign shouldn't fail other than out of memory, so I don't think falling back to std::align helps anyway.

kmillar commented on src/main/AllocatorSuperblock.cpp:
```diff
@@ -0,0 +1,307 @@
+#include "rho/AllocatorSuperblock.hpp"
+
+// Arena for small-object superblocks:
+uintptr_t arena_superblock_start = 0;
```
Can these be made file-local variables instead of globals?

llbit replied:
Yes, they can.

kmillar commented on src/main/AllocatorSuperblock.cpp:
```diff
@@ -0,0 +1,307 @@
+      + num_superblock * s_small_superblock_size;
+  arena_superblock_next = arena_superblock_start;
+
+#ifdef HAVE_ADDRESS_SANITIZER
```
There's no need for the #ifdef statements here.  ASAN_POISON_MEMORY_REGION is a no-op if asan isn't enabled.

llbit replied:
Oh, that removes the need for several of these `#ifdef` blocks.

ArunChauhan commented on src/main/GCNodeAllocator.cpp:
You don't need the "static" keyword for functions in the anonymous namespace.

kmillar commented on src/main/GCNodeAllocator.cpp:
IIUC, you're rounding up the size of each of the largest allocations to the next power of two.  What impact does that have?  Might it be more problematic on some OSes or configurations than others?

kmillar commented on src/main/GCNodeAllocator.cpp:
IIUC, the entries on these large allocation freelists are never released.

That'd probably be OK if we added calls to madvise with MADV_FREE or MADV_DONTNEED (the availability and semantics of these differs between different OSes, so this is a bit tricky).  posix_madvise doesn't have an option with the right semantics to be useful here.

Alternatively, some heuristic for dropping unused allocations could be used.

llbit commented on src/main/GCNodeAllocator.cpp:
It does cause some internal fragmentation and higher average memory usage, but it also improves allocation reuse - if a 4k allocation is freed, and then a 3k object needs to be allocated then the 4k allocation can be reused, whereas with BDWGC it would likely make a new allocation.

For an allocator that never gives back memory to the OS I think it's more important in the long run to be able to reuse more allocations.

llbit commented on src/main/GCNodeAllocator.cpp:
A maximum limit for the number of active freelist nodes could be added and then when it overflows the oldest freelist entries could be released.

Alternatively it would be possible to have a limit on the number of free byte in all freelists, and when the limit is reached all free nodes could be released.

llbit commented on src/main/GCNodeAllocator.cpp:
The extra internal fragmentation also affects the CPU caches, but for the benchmarks this seems to have very minimal effect if any - in a previous version of the allocator I just rounded up to the next 8-multiple, and the benchmark performance was not noticeably different.

kmillar commented on src/main/GCNodeAllocator.cpp:
The higher memory use concerns me -- these allocations can be on the order of ~ GB (large matrices and long vectors), so it could be a lot.
